netutil/netutil.go

```diff
@@ -54,3 +57,166 @@ func privateNetworkInterfaces(all []net.Interface, fallback []string, logger log
+
+	for _, name := range names {
+		addrs, err := interfaceAddrsFunc(name)
+		if err != nil {
+			level.Warn(logger).Log("msg", "error getting addresses for interface", "inf", name, "err", err)
+			continue
+		}
+
+		if len(addrs) <= 0 {
+			level.Warn(logger).Log("msg", "no addresses found for interface", "inf", name, "err", err)
+			continue
+		}
+		if ip := filterBestIP(addrs, enableInet6); ip.IsValid() {
+			// Select the best between what we've received
+			ipAddr = filterBestIP([]netip.Addr{ip, ipAddr}, enableInet6)
+		}
+		level.Debug(logger).Log("msg", "detected highest quality address", "ipAddr", ipAddr.String(), "inf", name)
+		if ipAddr.IsLinkLocalUnicast() || !ipAddr.IsValid() {
+			level.Debug(logger).Log("msg", "ignoring address", "ipAddr", ipAddr.String(), "inf", name)
+			continue
+		}
+
+		if enableInet6 {
+			if ipAddr.Is6() {
+				return ipAddr.String(), nil
+			}
+			continue
+		}
+
+		return ipAddr.String(), nil
+	}
+
+	level.Debug(logger).Log("msg", "detected IP address after looking up all configured interface names", "ipAddr", ipAddr.String())
+	if !ipAddr.IsValid() {
+		return "", fmt.Errorf("no useable address found for interfaces %s", names)
+	}
+	if ipAddr.IsLinkLocalUnicast() {
+		level.Warn(logger).Log("msg", "using link-local address", "address", ipAddr.String())
+	}
+	return ipAddr.String(), nil
+}
+
+// getInterfaceAddresses is the standard approach to collecting []net.Addr from a network interface by name.
+func getInterfaceAddresses(name string) ([]netip.Addr, error) {
+	inf, err := net.InterfaceByName(name)
+	if err != nil {
+		return nil, err
+	}
+
+	addrs, err := inf.Addrs()
+	if err != nil {
+		return nil, err
+	}
+
+	// Using netip.Addr to allow for easier and consistent address parsing.
+	// Without this, the net.ParseCIDR() that we might like to use in a test does
+	// not have the same net.Addr implementation that we get from calling
+	// interface.Addrs() as above.  Here we normalize on netip.Addr.
+	netaddrs := make([]netip.Addr, len(addrs))
+	for i, a := range addrs {
+		prefix, err := netip.ParsePrefix(a.String())
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to parse netip.Prefix")
+		}
+		netaddrs[i] = prefix.Addr()
+	}
+
+	return netaddrs, nil
+}
+
+// filterBestIP returns an opinionated "best" address from a list of addresses.
+// A high quality address is one that is considered routable, and not in the link-local address space.
+// A low quality address is a link-local address.
+// When an IPv6 is enabled using enableInet6, an IPv6 will be preferred over an equivalent quality IPv4 address,
+// otherwise IPv6 addresses are guaranteed to not be returned from this function.
+// Loopback addresses are never selected.
+func filterBestIP(addrs []netip.Addr, enableInet6 bool) netip.Addr {
+	var invalid, inetAddr, inet6Addr netip.Addr
+
+	for _, addr := range addrs {
+		if addr.IsLoopback() {
```

Comment:
You can move the `if addr.IsValid() {` directly here:

```suggestion
		if addr.IsLoopback() || !addr.IsValid() {
```